clang/test/CodeGenSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,167 @@
+
+        lambda_in_dependent_function<decltype(x)>();
+        // CHECK: call spir_func void @_Z28lambda_in_dependent_functionIZZ4mainENKUlvE_clEvEUlvE_Evv
+
+        lambda_no_dep<int, double>(3, 5.5);
+        // CHECK: call spir_func void @_Z13lambda_no_depIidEvT_T0_(i32 3, double 5.500000e+00)
+
+        int a = 5;
+        double b = 10.7;
+        auto y = [](int a) { return a; };
+        auto z = [](double b) { return b; };
+        lambda_two_dep<decltype(y), decltype(z)>();
+        // CHECK: store i32 5, i32* %a, align 4
+        // CHECK: store double 1.070000e+01, double* %b, align 8
+        // CHECK: call spir_func void @_Z14lambda_two_depIZZ4mainENKUlvE_clEvEUliE_ZZ4mainENKS0_clEvEUldE_Evv
+
+        lambda_two_dep<decltype(z), decltype(y)>();
+        // CHECK: call spir_func void @_Z14lambda_two_depIZZ4mainENKUlvE_clEvEUldE_ZZ4mainENKS0_clEvEUliE_Evv
+      });
+
+  kernel_single_task<class kernel2>(func<Derp>);
+
+  auto l1 = []() { return 1; };
+  auto l2 = [](decltype(l1) *l = nullptr) { return 2; };
+  kernel_single_task<class kernel3>(l2);
+  puts(__builtin_unique_stable_name(l2));
+
+
+  // CHECK: define linkonce_odr spir_func void @_Z14template_paramIiEvv
+  // CHECK: call spir_func void @puts(i8* getelementptr inbounds ([[INT1_SIZE]], [[INT1_SIZE]]* @[[INT1]]
+
+  // CHECK: define internal spir_func void @_Z14template_paramIZZ4mainENKUlvE_clEvEUlvE_Evv
+  // CHECK: call spir_func void @puts(i8* getelementptr inbounds ([[LAMBDA_SIZE]], [[LAMBDA_SIZE]]* @[[LAMBDA]]
+
+  // CHECK: define linkonce_odr spir_func void @_Z28lambda_in_dependent_functionIiEvv
+  // CHECK: call spir_func void @puts(i8* getelementptr inbounds ([[DEP_INT_SIZE]], [[DEP_INT_SIZE]]* @[[LAMBDA_IN_DEP_INT]]
+
+  // CHECK: define internal spir_func void @_Z28lambda_in_dependent_functionIZZ4mainENKUlvE_clEvEUlvE_Evv
+  // CHECK: call spir_func void @puts(i8* getelementptr inbounds ([[DEP_LAMBDA_SIZE]], [[DEP_LAMBDA_SIZE]]* @[[LAMBDA_IN_DEP_X]]
+
+  // CHECK: define linkonce_odr spir_func void @_Z13lambda_no_depIidEvT_T0_(i32 %a, double %b)
+  // CHECK: call spir_func void @puts(i8* getelementptr inbounds ([[NO_DEP_LAMBDA_SIZE]], [[NO_DEP_LAMBDA_SIZE]]* @[[LAMBDA_NO_DEP]]
+
+  // CHECK: define internal spir_func void @_Z14lambda_two_depIZZ4mainENKUlvE_clEvEUliE_ZZ4mainENKS0_clEvEUldE_Evv
+  // CHECK: call spir_func void @puts(i8* getelementptr inbounds ([[DEP_LAMBDA1_SIZE]], [[DEP_LAMBDA1_SIZE]]* @[[LAMBDA_TWO_DEP]]
+
+  // CHECK: define internal spir_func void @_Z14lambda_two_depIZZ4mainENKUlvE_clEvEUldE_ZZ4mainENKS0_clEvEUliE_Evv
+  // CHECK: call spir_func void @puts(i8* getelementptr inbounds ([[DEP_LAMBDA2_SIZE]], [[DEP_LAMBDA2_SIZE]]* @[[LAMBDA_TWO_DEP2]]
+
+  constexpr const char str[] = "lalala";
+  static_assert(__builtin_strcmp(__builtin_unique_stable_name(str), "_ZTSA7_Kc\00") == 0, "unexpected mangling");
+
+  // FIXME: Warn about expression with side effects in unevaluated context
+  int i = 0;
+  puts(__builtin_unique_stable_name(i++));
```

Comment:
@AaronBallman : I thought the side-effects thing was only for VLAs?  This one shouldn't have a problem, right?  Either way, a check for a warning should still SEMA, not codegen.

Reply:
Wait! I thought a problem/warning is expected because i is not being incremented at all i.e., expression in unevaluated context? 
Currently emitted IR is 
  store i32 0, i32* %i, align 4
  call spir_func void @puts(i8* getelementptr inbounds ([6 x i8], [6 x i8]* @usn_str.1, i32 0, i32 0)) #3
I don't see an increment instruction anywhere.

Reply:
That is an "unevaluated operand", which occurs in a few places, such as `decltype` or `sizeof`.  The intent is that we care about the type of the expression, but don't want it to do anything, so things such as side-effects just don't happen. So you wouldn't expect it to generate ANY code.

HOWEVER, there is an exception in the case of `sizeof` for VLAs, or "Variable Length Arrays" (see wikipedia for a good article).  VLAs are created with the size at the time of execution, so their size becomes an execution-time expression.  The result is that the side-effects happen.

Aaron and I went back and forth on this at one point, whether the VLA case side-effect should happen.  Otherwise, it shouldn't happen.

Reply:
Got it. I will remove the comment about the warning.